File: javascript/leetcode/109. Convert Sorted List to Binary Search Tree.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {ListNode} head
 * @return {TreeNode}
 */
var sortedListToBST = function (head) {
  if (!head) return null;

  let mid = head;
  let fast = head.next;
  let before_mid = null;

  while (fast) {
    before_mid = mid;
    mid = mid.next;
    fast = fast.next ? fast.next.next : null;
  }

  if (before_mid) before_mid.next = null;

  let root = new TreeNode(mid.val);
  if (Object.is(head, mid)) return root;

  root.left = sortedListToBST(head);
  root.right = sortedListToBST(mid.next);

  return root;
};

File: c++/leetcode/109. Convert Sorted List to Binary Search Tree.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution
{
public:
  TreeNode *sortedListToBST(ListNode *head)
  {
    if (!head)
      return nullptr;

    auto mid = head;
    auto fast = head->next;
    ListNode *before_mid = nullptr;

    // find middle of linked list
    while (fast)
    {
      before_mid = mid;
      mid = mid->next;
      fast = fast->next ? fast->next->next : nullptr;
    }

    // splits linked list into left and right parts
    if (before_mid)
      before_mid->next = nullptr;

    auto root = new TreeNode(mid->val);
    if (head == mid)
      return root;

    auto right = mid->next;
    auto left = head;

    root->left = sortedListToBST(left);
    root->right = sortedListToBST(right);

    return root;
  }
};